cmd/kubeadm/app/apis/output/types.go

```diff
@@ -32,3 +32,15 @@ type BootstrapToken struct {
 
 	kubeadmapiv1beta2.BootstrapToken
 }
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// UpgradePlanItem item represents information about upgrade plan for one component
+type UpgradePlanItem struct {
```

Comment:
i didn't do it for two reasons:
- it adds one level of complexity without adding any new information.
- it makes it impossible to use unbuffered output

Reply:
I don't think, that the lack of unbuffered output for upgrade plan is a big deal. The problem is, that we end up with a bunch of different YAML documents in a single output and that can be inconvenient to parse.

Reply:
I'd say they're pretty straightforward to parse. Here is an example: https://play.golang.com/p/BFCyyx8lx4R

Thank you for this. Playing with parsing helped me to understand that we'll probably need to add `yaml:"<name>"` tags to the public structures.

Reply:
for me, buffering makes sense only when there is considerable network delay between fetching the items. if "plan" does not have delay between printing the same items currently (i don't think it does), we can block until we have all items and have them better structured as @rosti suggests here:
https://github.com/kubernetes/kubernetes/pull/83941/files#r336393656

another problem with buffering here - what about if we want to have global fields for the "plan" output in the next version (v1alpha2) and not only a list of items? this will demand as separate struct for those...where the items are buffered and at the end we get these global fields in a separate strcture.

so, i think i'm leaning towards a +1 for https://github.com/kubernetes/kubernetes/pull/83941/files#r336393656
even if we sacrifice buffering.

yet, we are probably going to have the same discussion for other commands?

Reply:
I'll look what I can do here. It'd complicates the code it seems, but hopefully doable. I hope it's ok to make text output buffered too just for the sake of less code/consistency?

Reply:
i seem to be talking without recollecting how "upgrade plan" actually works.

```
[preflight] Running pre-flight checks.
[upgrade] Making sure the cluster is healthy:
[upgrade/config] Making sure the configuration is correct:
[upgrade/config] Reading configuration from the cluster...
[upgrade/config] FYI: You can look at this config file with 'kubectl -n kube-system get cm kubeadm-config -oyaml'
[upgrade] Fetching available versions to upgrade to
[upgrade/versions] Cluster version: v1.14.0
[upgrade/versions] kubeadm version: v1.14.0
...
table goes here
```

these are steps that require actions over a network...

so a problem here is the following, and i believe i spoke about this at some point.
we don't have good separation of structured output and random log entries like the above.

klog always prints to stderr, but kubeadm's sparse println calls print to stdout.
if "text" is used as the output format we don't care, but if a structured output is used, to me it makes sense the following:

- switch to print all log lines to stderr
- print the output the user cares about to stdout

this provides:
- the users can filter stdout and only care about it.
- based on exit status they can figure out if the command failed
- look at logs from stderr
- make the log output unbufferred and buffer only the structured output?

these aspects are missing from the kep...
@bart0sh we can discuss this in the kubeadm office hours if you want.

Reply:
that said, this is obviously a big change and we don't want to do it....
> switch to print all log lines to stderr

so a question here is do we want to go with the "let's buffer everything".
the problem with that is that the user will get no indication if a network call blocks for longer periods of time...

Reply:
So far I'm taking care of this step by step. For every output that I switch to structured/text I'm trying to make sure the output is not messed up with the prints. @fabriziopandini pointed out in [his comment above](https://github.com/kubernetes/kubernetes/pull/83941#pullrequestreview-304464424) that Print calls in the API should be considered as a technical debt and removing them is another good thing that's triggered by structured output implementation.